File: packages/page/src/query-rows.ts
import {
  hasQueryFieldMapping,
  type FieldValue,
  type QueryDataSourceFieldDefinition,
  type QueryFieldDefinition,
  type QueryRecordListFieldDefinition
} from './field';
import {
  fieldContractViolations,
  type FieldContractIssue
} from './result-field-contract';
import type { Row } from './snapshot';

/**
 * 查询结果归一化的问题契约:查询字段映射问题以 DQE 输出字段名定位,
 * 契约违规复用结果字段契约校验 Module 的分类并附映射上下文。
 * 与共享校验一致,不回显业务字段值。
 */
export type QueryRowNormalizationIssue =
  | { code: 'ROWS_NOT_ARRAY' }
  | { code: 'ROW_NOT_OBJECT'; rowIndex: number }
  | {
      code: 'MISSING_QUERY_FIELD';
      rowIndex: number;
      fieldId: string;
      queryField: string;
      actualFields: string[];
    }
  | {
      code: 'MISSING_DETAIL_QUERY_FIELD';
      rowIndex: number;
      fieldId: string;
      queryField: string;
      itemIndex: number;
      itemFieldId: string;
      itemQueryField: string;
      actualFields: string[];
    }
  | (FieldContractIssue & { queryField: string; itemQueryField?: string });

export type QueryRowsNormalizationResult =
  | { ok: true; rows: Row[]; issues: [] }
  | { ok: false; issues: QueryRowNormalizationIssue[] };

/**
 * 使用查询字段映射把 DQE 原始结果归一化为稳定页面字段。
 * PageDocument 的内嵌初始行与数据网关的动态响应共用此纯计算接缝;
 * 本模块只负责查询字段映射,字段类型、nullable 与日期时间规则
 * 全部委托结果字段契约校验 Module,不自持判断。
 */
export function normalizeQueryRows(
  value: unknown,
  fieldMappings: Record<string, QueryDataSourceFieldDefinition>
): QueryRowsNormalizationResult {
  if (!Array.isArray(value)) {
    return { ok: false, issues: [{ code: 'ROWS_NOT_ARRAY' }] };
  }

  const issues: QueryRowNormalizationIssue[] = [];
  const rows: Row[] = [];
  value.forEach((rawRow, rowIndex) => {
    if (!isRecord(rawRow)) {
      issues.push({ code: 'ROW_NOT_OBJECT', rowIndex });
      return;
    }

    const row: Row = {};
    for (const [fieldId, mapping] of Object.entries(fieldMappings)) {
      // 计算阶段产出字段不来自外部响应,归一化阶段跳过;它们由算子写入。
      if (!hasQueryFieldMapping(mapping)) continue;
      if (!Object.hasOwn(rawRow, mapping.queryField)) {
        issues.push({
          code: 'MISSING_QUERY_FIELD',
          rowIndex,
          fieldId,
          queryField: mapping.queryField,
          actualFields: Object.keys(rawRow)
        });
        continue;
      }

      const mapped =
        mapping.type === 'recordList'
          ? mapDetailItems(rawRow[mapping.queryField], mapping, rowIndex, fieldId)
          : { value: rawRow[mapping.queryField], issues: [], misses: NO_MISSES };
      issues.push(...mapped.issues);

      const violations = fieldContractViolations(mapped.value, mapping).filter(
        (violation) =>
          violation.code !== 'DETAIL_MISSING_FIELD' ||
          !mapped.misses.has(`${violation.itemIndex}:${violation.itemFieldId}`)
      );
      if (violations.length === 0 && mapped.issues.length === 0) {
        row[fieldId] = mapped.value as FieldValue;
      }
      issues.push(
        ...violations.map((violation) =>
          withQueryFields({ ...violation, rowIndex, fieldId }, mapping)
        )
      );
    }
    rows.push(row);
  });

  return issues.length === 0
    ? { ok: true, rows, issues: [] }
    : { ok: false, issues };
}

const NO_MISSES: ReadonlySet<string> = new Set();

interface MappedDetailValue {
  value: unknown;
  issues: QueryRowNormalizationIssue[];
  /** 已作为映射缺失上报的 `${itemIndex}:${itemFieldId}`,契约判定不再重复上报。 */
  misses: ReadonlySet<string>;
}

/**
 * 把嵌套明细的项按项级查询字段映射改写为稳定项字段;未映射的 DQE
 * 追加字段就地丢弃,非对象项原样透传交契约判定裁决。
 */
function mapDetailItems(
  value: unknown,
  mapping: QueryRecordListFieldDefinition,
  rowIndex: number,
  fieldId: string
): MappedDetailValue {
  if (!Array.isArray(value)) return { value, issues: [], misses: NO_MISSES };
  const issues: QueryRowNormalizationIssue[] = [];
  const misses = new Set<string>();
  const items = value.map((rawItem, itemIndex) => {
    if (!isRecord(rawItem)) return rawItem;
    const record: Record<string, unknown> = {};
    for (const [itemFieldId, itemMapping] of Object.entries(mapping.items.fields)) {
      if (!Object.hasOwn(rawItem, itemMapping.queryField)) {
        misses.add(`${itemIndex}:${itemFieldId}`);
        issues.push({
          code: 'MISSING_DETAIL_QUERY_FIELD',
          rowIndex,
          fieldId,
          queryField: mapping.queryField,
          itemIndex,
          itemFieldId,
          itemQueryField: itemMapping.queryField,
          actualFields: Object.keys(rawItem)
        });
        continue;
      }
      record[itemFieldId] = rawItem[itemMapping.queryField];
    }
    return record;
  });
  return { value: items, issues, misses };
}

function withQueryFields(
  issue: FieldContractIssue,
  mapping: QueryFieldDefinition
): QueryRowNormalizationIssue {
  const itemQueryField =
    'itemFieldId' in issue && mapping.type === 'recordList'
      ? mapping.items.fields[issue.itemFieldId]?.queryField
      : undefined;
  return {
    ...issue,
    queryField: mapping.queryField,
    ...(itemQueryField === undefined ? {} : { itemQueryField })
  };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

File: metriccanvas-page-assets/page-assets-service/src/main/java/com/huawei/cdi/pageassets/domain/page/document/QueryRows.java
package com.huawei.cdi.pageassets.domain.page.document;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.huawei.cdi.pageassets.domain.page.json.Json;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * 使用查询字段映射把 DQE 原始结果归一化为稳定页面字段。字段类型、nullable 与日期时间规则
 * 全部委托 {@link FieldContract}，本类只负责映射。
 */
public final class QueryRows {
    private QueryRows() {
    }

    /**
     * code 取值：ROWS_NOT_ARRAY / ROW_NOT_OBJECT / MISSING_QUERY_FIELD / MISSING_DETAIL_QUERY_FIELD，
     * 或 {@link FieldContract.Violation} 的分类并附映射上下文（queryField / itemQueryField）。
     */
    public record Issue(String code, Integer rowIndex, String fieldId, String queryField, Integer itemIndex,
                        String itemFieldId, String itemQueryField, FieldContract.Violation violation) {
    }

    public record Result(ArrayNode rows, List<Issue> issues) {
        public Result {
            issues = List.copyOf(issues);
        }

        public boolean ok() {
            return issues.isEmpty();
        }
    }

    public static Result normalize(JsonNode value, JsonNode fieldMappings) {
        List<Issue> issues = new ArrayList<>();
        ArrayNode rows = JsonNodeFactory.instance.arrayNode();
        if (!Json.isArray(value)) {
            issues.add(new Issue("ROWS_NOT_ARRAY", null, null, null, null, null, null, null));
            return new Result(rows, issues);
        }
        for (int rowIndex = 0; rowIndex < value.size(); rowIndex++) {
            JsonNode rawRow = value.get(rowIndex);
            if (!Json.isRecord(rawRow)) {
                issues.add(new Issue("ROW_NOT_OBJECT", rowIndex, null, null, null, null, null, null));
                continue;
            }
            ObjectNode row = JsonNodeFactory.instance.objectNode();
            for (Map.Entry<String, JsonNode> entry : Json.entries(fieldMappings)) {
                String fieldId = entry.getKey();
                JsonNode mapping = entry.getValue();
                if (!DataSources.hasQueryFieldMapping(mapping)) {
                    continue;
                }
                String queryField = DataSources.queryField(mapping);
                if (!rawRow.has(queryField)) {
                    issues.add(new Issue("MISSING_QUERY_FIELD", rowIndex, fieldId, queryField, null, null, null, null));
                    continue;
                }
                JsonNode raw = rawRow.get(queryField);
                Mapped mapped = "recordList".equals(DataSources.type(mapping))
                        ? mapDetailItems(raw, mapping, rowIndex, fieldId)
                        : new Mapped(raw, List.of(), Set.of());
                issues.addAll(mapped.issues);

                List<FieldContract.Violation> violations = new ArrayList<>();
                for (FieldContract.Violation violation : FieldContract.violations(mapped.value, mapping)) {
                    if ("DETAIL_MISSING_FIELD".equals(violation.code())
                            && mapped.misses.contains(violation.itemIndex() + ":" + violation.itemFieldId())) {
                        continue;
                    }
                    violations.add(violation);
                }
                if (violations.isEmpty() && mapped.issues.isEmpty()) {
                    row.set(fieldId, mapped.value);
                }
                for (FieldContract.Violation violation : violations) {
                    String itemQueryField = null;
                    if (violation.itemFieldId() != null && "recordList".equals(DataSources.type(mapping))) {
                        JsonNode itemMapping = Json.get(Json.get(Json.get(mapping, "items"), "fields"), violation.itemFieldId());
                        itemQueryField = DataSources.queryField(itemMapping);
                    }
                    issues.add(new Issue(violation.code(), rowIndex, fieldId, queryField, violation.itemIndex(),
                            violation.itemFieldId(), itemQueryField, violation));
                }
            }
            rows.add(row);
        }
        return new Result(rows, issues);
    }

    private record Mapped(JsonNode value, List<Issue> issues, Set<String> misses) {
    }

    /** 把嵌套明细的项按项级查询字段映射改写为稳定项字段；未映射的 DQE 追加字段就地丢弃。 */
    private static Mapped mapDetailItems(JsonNode value, JsonNode mapping, int rowIndex, String fieldId) {
        if (!Json.isArray(value)) {
            return new Mapped(value, List.of(), Set.of());
        }
        List<Issue> issues = new ArrayList<>();
        Set<String> misses = new HashSet<>();
        ArrayNode items = JsonNodeFactory.instance.arrayNode();
        JsonNode itemMappings = Json.get(Json.get(mapping, "items"), "fields");
        String queryField = DataSources.queryField(mapping);
        for (int itemIndex = 0; itemIndex < value.size(); itemIndex++) {
            JsonNode rawItem = value.get(itemIndex);
            if (!Json.isRecord(rawItem)) {
                items.add(rawItem);
                continue;
            }
            ObjectNode record = JsonNodeFactory.instance.objectNode();
            for (Map.Entry<String, JsonNode> entry : Json.entries(itemMappings)) {
                String itemFieldId = entry.getKey();
                String itemQueryField = DataSources.queryField(entry.getValue());
                if (!rawItem.has(itemQueryField)) {
                    misses.add(itemIndex + ":" + itemFieldId);
                    issues.add(new Issue("MISSING_DETAIL_QUERY_FIELD", rowIndex, fieldId, queryField, itemIndex,
                            itemFieldId, itemQueryField, null));
                    continue;
                }
                record.set(itemFieldId, rawItem.get(itemQueryField));
            }
            items.add(record);
        }
        return new Mapped(items, issues, misses);
    }
}
